sklearn/tree/_tree.pyx

```diff
@@ -1056,6 +1056,28 @@ cdef class Tree:
 
         return out
 
+    cpdef compute_node_depths(self):
+        """Compute the depth of each node in a tree.
+
+        Returns
+        -------
+        depths : ndarray of shape (self.node_count,), dtype=np.int64
+            The depth of each node in the tree.
+
+        """
+        cdef cnp.int64_t[::1] depths = np.empty(self.node_count, dtype=np.int64)
```

Comment:
Without the following, I don't think that we will get the proper speed-up because we will have too many look up in the array.

```suggestion
        cdef cnp.int64_t[::1] depths = np.empty(self.node_count, dtype=np.int64)
        cnp.npy_intp[:] children_left = self.children_left
        cnp.npy_intp[:] children_right = self.children_right
```